src/lib/sql/sql_pipeline.cpp

```diff
@@ -0,0 +1,183 @@
+    throw std::runtime_error("Error while compiling query plan:\n  " + std::string(exception.what()));
+  }
+
+  return _unopt_logical_plan;
+}
+
+const std::vector<std::shared_ptr<AbstractLQPNode>>& SQLPipeline::get_optimized_logical_plan() {
+  if (!_opt_logical_plan.empty()) {
+    return _opt_logical_plan;
+  }
+
+  auto unopt_lqp = get_unoptimized_logical_plan();
+  _opt_logical_plan.reserve(unopt_lqp.size());
+
+  try {
+    for (const auto& node : unopt_lqp) {
+      _opt_logical_plan.push_back(Optimizer::get().optimize(node));
+    }
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while optimizing query plan:\n  " + std::string(exception.what()));
+  }
+
+  return _opt_logical_plan;
+}
+
+const SQLQueryPlan& SQLPipeline::get_query_plan() {
+  if (_query_plan) {
+    return *(_query_plan.get());
+  }
+
+  const auto& lqp_roots = get_optimized_logical_plan();
+  auto plan = std::make_unique<SQLQueryPlan>();
+
+  auto started = std::chrono::high_resolution_clock::now();
+
+  try {
+    for (const auto& node : lqp_roots) {
+      auto optimized = Optimizer::get().optimize(node);
+      auto op = LQPTranslator{}.translate_node(optimized);
+      plan->add_tree_by_root(op);
+    }
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while translating query plan:\n  " + std::string(exception.what()));
+  }
+
+  if (_use_mvcc) {
+    plan->set_transaction_context(_transaction_context);
+  }
+
+  auto done = std::chrono::high_resolution_clock::now();
+  _compile_time_sec = std::chrono::duration<double>(done - started).count();
+
+  _query_plan = std::move(plan);
+  return *(_query_plan.get());
+}
+
+const std::vector<std::shared_ptr<OperatorTask>>& SQLPipeline::get_tasks() {
+  if (!_op_tasks.empty()) {
+    return _op_tasks;
+  }
+
+  const auto& query_plan = get_query_plan();
+  try {
+    _op_tasks = query_plan.create_tasks();
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while creating tasks:\n  " + std::string(exception.what()));
+  }
+
+  return _op_tasks;
+}
+
+const std::shared_ptr<const Table>& SQLPipeline::get_result_table() {
+  if (_result_table) {
+    return _result_table;
+  }
+
+  DebugAssert(_use_mvcc, "Cannot execute query without a transaction context.");
+
+  const auto& op_tasks = get_tasks();
+
+  auto started = std::chrono::high_resolution_clock::now();
+
+  try {
+    for (const auto& task : op_tasks) {
+      task->schedule();
+    }
+  } catch (const std::exception& exception) {
+    _transaction_context->rollback();
+    throw std::runtime_error("Error while executing tasks:\n  " + std::string(exception.what()));
+  }
+
+  if (_auto_commit) {
+    _transaction_context->commit();
+  }
+
+  auto done = std::chrono::high_resolution_clock::now();
+  _execution_time_sec = std::chrono::duration<double>(done - started).count();
+
+  _result_table = op_tasks.back()->get_operator()->get_output();
```

Comment:
You're not waiting for the tasks to be finished, are you?

Reply:
This is a problem of the current console as well. Have a look at `CurrentScheduler::schedule_and_wait_for_tasks`.